scalafix-cli/src/main/scala/scalafix/internal/interfaces/ScalafixArgumentsImpl.scala

```diff
@@ -76,7 +80,48 @@ final case class ScalafixArgumentsImpl(args: Args = Args.default)
       customDependenciesCoordinates,
       Versions.scalaVersion
     )
-    val extraURLs = customURLs.asScala ++ customDependenciesJARs.asScala
+
+    // External rules are built against `scalafix-core` to expose `scalafix.v1.Rule` implementations. The
+    // classloader loading `scalafix-cli` already contains  `scalafix-core` to be able to discover them (which
+    // is why it must be the parent of the one loading the tool classpath), so effectively, the version/instance
+    // in the tool classpath will not be used. This adds a sanity check to warn or prevent the user in case of
+    // mismatch.
+    val scalafixCore = coursierapi.Module.parse(
+      "ch.epfl.scala::scalafix-core",
+      coursierapi.ScalaVersion.of(Versions.scalaVersion)
+    )
+    customDependenciesJARs.getDependencies.asScala
+      .find(_.getModule == scalafixCore)
+      .foreach { dependency =>
+        // We only check compatibility against THE scalafix-core returned by the coursier resolution, but
+        // this is not exhaustive as some old rules might coexist with recent ones, so the older versions
+        // get evicted. coursier-interface does not provide more granularity while the native coursier
+        // is stuck on a 2-years old version because it no longer cross-publishes for 2.11, so for now,
+        // the easiest option would be to run several resolutions in isolation, which seems like a massive
+        // cost for just issuing a warning.
+        Compatibility.pvp(dependency.getVersion, Versions.nightly) match {
+          case Incompatible =>
+            throw new ScalafixException(
+              s"Scalafix version ${Versions.scalaVersion} cannot load the registered external rules, " +
+                s"please upgrade to ${dependency.getVersion} or later"
+            )
+          case Unknown =>
+            args.out.println(
```

Comment:
sbt-scalafix sets the printstream [very late](https://github.com/scalacenter/sbt-scalafix/blob/0a4dfc6a32d67b701235876fe5658f04c9c9687a/src/main/scala/scalafix/sbt/ScalafixPlugin.scala#L390), so this gets dumped on plain stdout. It would be possible to wire a fallback sbt logger from the beginning.